Mathlib/FieldTheory/Isaacs.lean

```diff
@@ -0,0 +1,116 @@
+/-
+Copyright (c) 2024 Junyan Xu. All rights reserved.
+Released under Apache 2.0 license as described in the file LICENSE.
+Authors: Junyan Xu
+-/
+import Mathlib.FieldTheory.PrimitiveElement
+import Mathlib.GroupTheory.CosetCover
+
+/-!
+# Characterize algebraic extensions by minimal polynomials
+
+## Main results
+
+`Field.nonempty_algHom_of_exist_roots` says if `E/F` and `K/F` are field extensions
+with `E/F` algebraic, and if the minimal polynomial of every element of `E` over `F` has a root
+in `K`, then there exists an `F`-embedding of `E` into `K`. If `E/F` and `K/F` have the same
+set of minimal polynomials, then `E` and `K` are isomorphic as `F`-algebras. As a corollary:
+
+`IsAlgClosure.of_exist_roots`: if `E/F` is algebraic and every monic irreducible polynomial
+in `F[X]` has a root in `E`, then `E` is an algebraic closure of `F`.
+
+## Reference
+
+[Isaacs1980] *Roots of Polynomials in Algebraic Extensions of Fields*,
+The American Mathematical Monthly
+
+-/
+
+namespace Field
+
+open Polynomial IntermediateField
+
+variable {F E K : Type*} [Field F] [Field E] [Field K] [Algebra F E] [Algebra F K]
+variable [alg : Algebra.IsAlgebraic F E]
+
+theorem nonempty_algHom_of_exist_roots (h : ∀ x : E, ∃ y : K, aeval y (minpoly F x) = 0) :
+    Nonempty (E →ₐ[F] K) := by
+  refine nonempty_algHom_of_exist_lifts_finset
+    fun S ↦ ⟨⟨adjoin F S, Nonempty.some ?_⟩, subset_adjoin _ _⟩
+  let p := (S.prod <| minpoly F).map (algebraMap F K)
+  let K' := SplittingField p
+  have splits s (hs : s ∈ S) : (minpoly F s).Splits (algebraMap F K') := by
+    apply splits_of_splits_of_dvd _
+      (Finset.prod_ne_zero_iff.mpr fun _ _ ↦ minpoly.ne_zero <| (alg.isIntegral).1 _)
+      ((splits_map_iff _ _).mp <| SplittingField.splits p) (Finset.dvd_prod_of_mem _ hs)
+  let K₀ := (⊥ : IntermediateField K K').restrictScalars F
+  let FS := adjoin F (S : Set E)
+  let Ω := FS →ₐ[F] K'
+  have := finiteDimensional_adjoin (S := (S : Set E)) fun _ _ ↦ (alg.isIntegral).1 _
+  let M (ω : Ω) := Subalgebra.toSubmodule (K₀.comap ω).toSubalgebra
+  have : ⋃ ω : Ω, ↑(M ω) = @Set.univ FS :=
+    Set.eq_univ_of_forall fun ⟨α, hα⟩ ↦ Set.mem_iUnion.mpr <| by
+      have ⟨β, hβ⟩ := h α
+      let ϕ : F⟮α⟯ →ₐ[F] K' := (IsScalarTower.toAlgHom _ _ _).comp ((AdjoinRoot.liftHom _ _ hβ).comp
+        (adjoinRootEquivAdjoin F <| (alg.isIntegral).1 _).symm.toAlgHom)
+      have ⟨ω, hω⟩ := exists_algHom_adjoin_of_splits
+        (fun s hs ↦ ⟨(alg.isIntegral).1 _, splits s hs⟩) ϕ (adjoin_simple_le_iff.mpr hα)
+      refine ⟨ω, β, ((DFunLike.congr_fun hω <| AdjoinSimple.gen F α).trans ?_).symm⟩
+      rw [AlgHom.comp_apply, AlgHom.comp_apply, AlgEquiv.coe_algHom,
+        adjoinRootEquivAdjoin_symm_apply_gen, AdjoinRoot.liftHom_root]
+      rfl
```

Comment:
I'm really happy that this works now. But I still don't understand why the timeout went away. Do we risk that it suddenly reappears? Can you please comment briefly on how you avoid it now?